From 66ac4f5c5d1511e16180c592dde1b8e4bf385c6f Mon Sep 17 00:00:00 2001
From: Derek Higgins <derekh@redhat.com>
Date: Tue, 23 Jul 2013 10:34:45 +0100
Subject: [PATCH] Ensure metadata is available for os-refresh-config

On Feodra 18 during boot when cron runs os-refresh-config, a race
condition was present, cron was getting run before the metadata was
retrieved. As a result in 49-heat-localip the loop that edited the
metadata was editing nothing.

Change-Id: I1b8048daf42786b9a9008db2062bd62b6a7ee7b9
---
 .../os-refresh-config/configure.d/49-heat-localip  |    7 +++++++
 .../os-refresh-config/install.d/75-cfn-hup-cronjob |   11 +++++++++++
 2 files changed, 18 insertions(+)

diff --git a/elements/heat-localip/os-refresh-config/configure.d/49-heat-localip b/elements/heat-localip/os-refresh-config/configure.d/49-heat-localip
index d5ccb8d..ee5dd39 100755
--- a/elements/heat-localip/os-refresh-config/configure.d/49-heat-localip
+++ b/elements/heat-localip/os-refresh-config/configure.d/49-heat-localip
@@ -8,6 +8,13 @@ if [ -z "$local_ip" ]; then
   # routes.
   exit 0
 fi
+
+# We may have to wait for metadata to appear if this script is being started on the first boot
+for i in {0..10} ; do
+    [ -e /var/lib/heat-cfntools/cfn-init-data -o -e /var/cache/heat-cfntools/last_metadata ] && break
+    sleep 2
+done
+
 for f in /var/lib/heat-cfntools/cfn-init-data /var/cache/heat-cfntools/last_metadata ; do
     if [ -e $f ] ; then
         sed -i "s/\<0\.0\.0\.0\>/$local_ip/g" $f
diff --git a/elements/os-refresh-config/install.d/75-cfn-hup-cronjob b/elements/os-refresh-config/install.d/75-cfn-hup-cronjob
index c5548fd..69beb08 100755
--- a/elements/os-refresh-config/install.d/75-cfn-hup-cronjob
+++ b/elements/os-refresh-config/install.d/75-cfn-hup-cronjob
@@ -10,3 +10,14 @@ PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin
 # This is a workaround for cfn-hup not actually being a daemon
 */5 * * * * root cfn-hup --no-daemon
 EOF
+
+# If using systemd in order of guarantee metadata is available,
+# ensure cron doesn't come up until after cloud-init has run
+if [ -e /usr/lib/systemd/system/cloud-final.service ] ; then
+    # We are editing a file packaged in cloud-init, so we update 
+    # cloud-init now, if its updated later it will undo our changes
+    yum update -y cloud-init || true
+    # TODO : this should be possible by placing a link in crond.service.wants
+    # but I couldn't get this to work
+    sed -i -e 's/Requires=cloud-config.target/Requires=cloud-config.target\nBefore=crond.service/g' /usr/lib/systemd/system/cloud-final.service
+fi
-- 
1.7.9.5

